compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -48,47 +52,67 @@ struct CollapseDimensionsPass
 
 /// Searches the same sequence in all the affine maps and collapses these
 /// dimensions. It only applies these to "parallel" loops without mixing them
-/// with "reduction" types.
+/// with "reduction" types. It is expected that the `genericOp` has projected
+/// permutations only as indexing maps. (Checked using `isEligibleForCollapse`).
 static SmallVector<ReassociationIndices>
 getCollapsibleLoops(linalg::GenericOp genericOp) {
   SmallVector<ReassociationIndices> contiguousLoops;
 
-  SmallVector<unsigned> pDims;
+  SmallVector<unsigned> pDims, rDims;
   genericOp.getParallelDims(pDims);
-  if (pDims.size() < 2)
-    return contiguousLoops;
-
-  llvm::SmallDenseSet<unsigned> pLoops(pDims.begin(), pDims.end());
+  genericOp.getReductionDims(rDims);
+  llvm::SmallDenseSet<unsigned> pDimsSet, rDimsSet;
+  pDimsSet.insert(pDims.begin(), pDims.end());
+  rDimsSet.insert(rDims.begin(), rDims.end());
 
   auto hasAllMapsSameSequence = [&](AffineExpr preExpr, AffineExpr nextExpr) {
+    // Either both `preExpr` and `nextExpr` are contiguous, or are missing in
+    // all the maps.
     for (AffineMap map : genericOp.getIndexingMapsArray()) {
-      bool foundSeq = false;
+      // If map has no results, no need to check.
+      if (map.getNumResults() == 0) {
+        continue;
+      }
       for (auto [index, resultExpr] : llvm::enumerate(map.getResults())) {
-        if (resultExpr == nextExpr) {
-          foundSeq = (index > 0 && preExpr == map.getResult(index - 1));
-          break;
+        // If we find the preExpr, we should find the nextExpr.
+        if (resultExpr == preExpr) {
+          if (index == map.getNumResults() - 1) {
+            // Reached end of list. Return false;
+            return false;
+          }
+          if (map.getResult(index + 1) != nextExpr) {
+            return false;
+          }
+        }
+        // If we find nextExpr at the beginning, then return false.
+        if (index == 0 && resultExpr == nextExpr) {
+          return false;
         }
       }
```

Comment:
Following up on the previous discussion, I thought this would work
```suggestion
# remove for (auto [index, resultExpr] : llvm::enumerate(map.getResults())) { too, github won't let me suggest across deleted lines.
        if (map.getNumResults() > 0 && map.getResult(0) == nextExpr) {
          return false;
        }
        std::optional<unsigned> maybeIndex = map.getResultPosition(preExpr);
        if (!maybeIndex) {
          continue;
        }
        if (*maybeIndex >= map.getNumResults()) {
          return false;
        }
        if (map.getResult(*maybeIndex + 1) != nextExpr) {
          return false;
        }
```

Reply:
Also, looking at the logic here again, I'm a little confused by `index == 0 && resultExpr == nextExpr`. Why do we only fail when `nextExpr` is alone if it is at `index == 0`?

Reply:
Hmmm, maybe... I tried just reuse the logic that existed. It will take some time for me to get back into the context and work out the details, but I am trying to understand why we cant just compare the expression directly. That effectively does the same thing?

Reply:
> Also, looking at the logic here again, I'm a little confused by `index == 0 && resultExpr == nextExpr`. Why do we only fail when `nextExpr` is alone if it is at `index == 0`?

basically you want to have either  `prevExpr` , `nextExpr` to be 
1) both present and consecutive, or
2) both absent.
If only `nextExpr` is at the start then that fails this conditions. I think after this, the following logic accounts for both of these. It looks for `prevExpr` and checks that the next expr is `nextExpr`.

Reply:
The loop as written was kind of hard for me to understand and required looking at how the helper was used to know what the valid `AffineExpr`s were for `preExpr` and `nextExpr`. Then it just looked like you were doing essentially the same thing that `getResultPosition` was doing. At a minimum it would probably help to add a short comment on that helper explaining exactly what it's trying to match.

Reply:
> > Also, looking at the logic here again, I'm a little confused by `index == 0 && resultExpr == nextExpr`. Why do we only fail when `nextExpr` is alone if it is at `index == 0`?
> 
> basically you want to have either `prevExpr` , `nextExpr` to be
> 
>     1. both present and consecutive, or
> 
>     2. both absent.
>        If only `nextExpr` is at the start then that fails this conditions. I think after this, the following logic accounts for both of these. It looks for `prevExpr` and checks that the next expr is `nextExpr`.

Ok, makes perfect sense. Then I claim that as written, something like `affine_map<(d0, d1, d2) -> (e0, e1, nextExpr)>` will pass through with `true` because `preExpr` is not present and thus it never checks for `nextExpr` outside of the first iteration. Is there a reason we can't expect to see maps like that?

Reply:
Ah, you are right. I'll fix the logic for your case. Reason we need both is that you have to check that these dimensions are contiguous everywhere then you can collapse these dimensions. If one is present and other is not anywhere, then if you collapse it you will have to use mods/divs in indexing maps (with potentially symbols) to recover the lone dimension.

Reply:
Fixed the issue @qedawkins PTAL. Also added some comments.